storage/scan_queue.go

```diff
@@ -145,42 +148,57 @@ func (sq *scanQueue) process(now proto.Timestamp, rng *Range) error {
 			RaftID:    rng.Desc.RaftID,
 		},
 	}
+	var oldestIntentNanos int64 = math.MaxInt64
+	var wg sync.WaitGroup
 	var expBaseKey proto.Key
 	var keys []proto.EncodedKey
 	var vals [][]byte
-	for ; iter.Valid(); iter.Next() {
-		baseKey, ts, isValue := engine.MVCCDecodeKey(iter.Key())
-		if !isValue {
-			// Moving to the next key (& values). If there's more than a
-			// single value for the key, possibly send for GC.
-			if len(keys) > 1 {
-				meta := &proto.MVCCMetadata{}
-				if err := gogoproto.Unmarshal(vals[0], meta); err != nil {
-					log.Errorf("unable to unmarshal MVCC metadata for key %q: %s", keys[0], err)
-				} else {
-					// In the event that there's an active intent, send for
-					// intent resolution if older than the threshold.
-					startIdx := 1
-					if meta.Txn != nil {
-						if meta.Txn.Timestamp.Less(intentExp) {
-							// TODO(spencer): push txn / resolve intent. Don't forget to
-							// update the oldest intent nanos if the resolve fails.
-						} else if scanMeta.OldestIntentNanos == nil || meta.Timestamp.WallTime < *scanMeta.OldestIntentNanos {
-							scanMeta.OldestIntentNanos = gogoproto.Int64(meta.Timestamp.WallTime)
-						}
-						// With an active intent, GC ignores MVCC metadata & intent value.
-						log.Infof("there's an intent on key %q", expBaseKey)
-						startIdx = 2
-					}
-					// See if any values may be GC'd.
-					if gcTS := gc.Filter(keys[startIdx:], vals[startIdx:]); !gcTS.Equal(proto.ZeroTimestamp) {
-						// TODO(spencer): need to split the requests up into
-						// multiple requests in the event that more than X keys
-						// are added to the request.
-						gcArgs.Keys = append(gcArgs.Keys, proto.InternalGCRequest_GCKey{Key: expBaseKey, Timestamp: gcTS})
+
+	// processKeysAndValues is invoked with each key and its set of
+	// values. Intents older than the intent age threshold are sent for
+	// resolution and values after the MVCC metadata, and possible
+	// intent, are sent for garbage collection.
+	processKeysAndValues := func() {
+		// If there's more than a single value for the key, possibly send for GC.
+		if len(keys) > 1 {
+			meta := &proto.MVCCMetadata{}
+			if err := gogoproto.Unmarshal(vals[0], meta); err != nil {
+				log.Errorf("unable to unmarshal MVCC metadata for key %q: %s", keys[0], err)
+			} else {
+				// In the event that there's an active intent, send for
+				// intent resolution if older than the threshold.
+				startIdx := 1
+				if meta.Txn != nil {
+					// Resolve intent asynchronously in a goroutine if the intent
+					// is older than the intent expiration threshold.
+					if meta.Timestamp.Less(intentExp) {
+						log.Infof("base=%q, now=%s, meta ts=%s, intentExp=%s", expBaseKey, now, meta.Timestamp, intentExp)
+						wg.Add(1)
+						go sq.resolveIntent(rng, expBaseKey, meta, &oldestIntentNanos, &wg)
+					} else if meta.Timestamp.WallTime < atomic.LoadInt64(&oldestIntentNanos) {
+						atomic.StoreInt64(&oldestIntentNanos, meta.Timestamp.WallTime)
```

Comment:
Another goroutine could write to oldestIntentNanos between the load and store. This should probably just use a mutex.